text/0034-colored-coins/0034-colored-coins.md

```diff
@@ -0,0 +1,359 @@
+      `index of the output in the transaction`.
+    - Mark the `color` of the output as `OutputID`.
+    - Book the output into the ledger.
+- **MOVE** opcode:
+    - Calculate `OutputID`, the unique identifier of the output in the UTXO DAG. `OutputID` = `TransactionID` ||
+      `index of the output in the transaction`.
+    - Book the output into the ledger.
+- **UNCOLOR** opcode:
+    - Calculate `OutputID`, the unique identifier of the output in the UTXO DAG. `OutputID` = `TransactionID` ||
+      `index of the output in the transaction`
+    - Mark the `color` of the output as `ColorIOTA`. <i>This might not be necessary, as unlocking an `UNCOLOR` output
+      adds `ColorIOTA` to the positive side of the balance sheet. With marking the booked output with `ColorIOTA`,
+      we double enforce this rule, but also lose the information on which color it uncolored. For that, one has to
+      load the transaction and observe the outputs in their original form. </i>
+    - Book the output into the ledger.
+
+### Notes
+- Outputs in the serialized transaction must be in lexicographic order, therefore there can be no two transactions that
+  define the same UTXO mutations while having different transaction IDs.
+- Due to the ordering, any `SigLockedColoredOutput` comes after any `SigLockedSingleOutput` within the transaction.
+- Due to the ordering, `MINT`, `MOVE` and `UNCOLOR` outputs are grouped together in the list of outputs, which helps to
+  identify the purpose of the transaction by looking at its layout.
+- One transaction might mint several coins with unique colors. The maximum amount of different colors that can be
+  minted in a single transaction is exactly the maximum allowed number of outputs, that is `128`.
+- One transaction might uncolor several colored coins. The maximum amount of different colors that can be uncolored in
+  a single transaction is exactly the maximum allowed number of outputs, that is `128`.
+- It is not possible to mint a color in a transaction and uncolor it in the same transaction, since uncoloring can only
+  happen if coins with that particular color are present in list of unlocked inputs.
+- The size of a `SigLockedColoredOutput` is exactly `76 bytes`, compared to the `42 bytes` size of a
+  `SigLockedSingleOutput`.
+- A `SigLockedColoredOutput` with `OPCODE=MOVE` and `Color=ColorIOTA` is functionally identical to a
+  `SigLockedSingleOutput`, but the latter is preferable due to its smaller size.
+
+### Dust protection
+
+Dust protection refers to preventing the unnecessary increase of the ledger database by splitting up funds into several,
+smaller outputs. Chrysalis Phase 2 introduced a new dust protection mechanism, that defines on a protocol level when
+such small outputs can be accepted.
+Dust protection doesn't change with the introduction of colored coins, a `SigLockedColoredOutput` is considered to be
```

Comment:
Really? same dust solution?

Wouldn't it be better for colored coins to force the issuer to deposit funds and lock them rather than the users?

Reply:
If the use case is utility tokens or security tokens, maybe yes. IMO the current chrysalis dust protection was designed with microtransactions in mind, and not thinkning about smart contract request tokens, NFTs and other fancy stuff we can do with colored coins. These different use cases introduce new requirements and constraints. At the end of the day, colored coins and smart contracts will increase the amount of load on nodes and dust in the DB, and there is no dust solution that can prevent this from happening, only control it to a certain extent.